Description: Fix build on armhf
 In particular, the hashing functions for strings (and similar
 classes) deal with overflowing signed integers, which is undefined
 behavior.  The last step of ANDing with 0x7fffffff to obtain a
 positive integer to use as an array index might be optimized out by
 the compiler.  We switch to unsigned integers instead.
Author: Dan Grayson <dan@math.uiuc.edu>
Author: Doug Torrance <dtorrance@piedmont.edu>
Bug: https://github.com/Macaulay2/M2/issues/1554
Last-Update: 2020-12-08

--- a/M2/Macaulay2/d/strings.d
+++ b/M2/Macaulay2/d/strings.d
@@ -36,9 +36,9 @@
 export (s:string) > (t:string) : bool := t<s;
 export (s:string) <= (t:string) : bool := !(t<s);
 export hash(s:string):int := (
-     h := 0;
-     foreach c in s do h = 31*h + int(c);
-     h & 0x7fffffff
+     h := uint(0);
+     foreach c in s do h = uint(31)*h + uint(c);
+     int(h & 0x7fffffff)
      );
 
 export index(s:string,offset:int):int := (
--- a/M2/Macaulay2/d/varnets.d
+++ b/M2/Macaulay2/d/varnets.d
@@ -63,14 +63,14 @@
      v.size = v.size + 1;
      v);
 export hash(v:vaNet):int := (
-     h := 323;
-     for i from 0 to v.size-1 do h = 41 * h + hash(v.nets.i);
-     h & 0x7fffffff);
+     h := uint(323);
+     for i from 0 to v.size-1 do h = uint(41) * h + uint(hash(v.nets.i));
+     int(h & 0x7fffffff));
 
 -- netfile
 
 export NetFile := {+ v:vaNet, x:varnet };
-export hash(n:NetFile):int := 0x7fffffff & (hash(n.v) + 43 * hash(n.x));
+export hash(n:NetFile):int := int(0x7fffffff & (uint(hash(n.v)) + uint(43) * uint(hash(n.x))));
 export newNetFile():NetFile := NetFile(newvaNet(),newvarnet());
 export (n:NetFile) << (i:int) : NetFile := ( n.x << i; n );
 export (n:NetFile) << (s:string) : NetFile := ( n.x << s; n );
--- a/M2/Macaulay2/d/varstrin.d
+++ b/M2/Macaulay2/d/varstrin.d
@@ -62,11 +62,11 @@
 -- export addstring(v:varstring,s:string):void := v<<s;
 
 export hash(s:varstring):int := (
-     h := 0;
+     h := uint(0);
      n := s.width;
      if n > length(s.str) then n = length(s.str);
-     for i from 0 to n-1 do h = 311*h + int(s.str.i);
-     h & 0x7fffffff );
+     for i from 0 to n-1 do h = uint(311)*h + uint(s.str.i);
+     int(h & 0x7fffffff) );
 export putdigit(o:varstring,x:int):void := o << (x + if x<10 then '0' else 'a'-10) ;
 
 export varstringarray := { a:array(string), n:int };
